### PR TITLE
Collect statistics from meters

### LINKED ISSUE #1573: Put STATS flag for all meters and collect this stats to OTSDB
http://flowgrammable.org/sdn/openflow/message-layer/metermod/

**NOTE: We assume that each meter has only one meterBand.**

There two types of meters on switch at this moment:
1. Meters for flows.
2. Meters for default rules.

**Metrics for flows:**

- pen.flow.meter.bits        - meterBand bits
- pen.flow.meter.bytes     - meterBand bytes
- pen.flow.meter.packets - meterBand packets

Tags: cookie, switchid, flowid, direction, meterid

**Metrics for default rule meters:**

- pen.switch.flow.system.meter.packets - meterBand packets
- pen.switch.flow.system.meter.bytes     - meterBand bytes
- pen.switch.flow.system.meter.bits        - meterBand bits

Tags: cookieHex, switchid, meterid

### PR DESCRIPTION
**Pull request by niksv**
Several metrics were added to OpenTSDB.

Metrics for flows:

- pen.flow.meter.bits - meterBand bits
- pen.flow.meter.bytes - meterBand bytes
- pen.flow.meter.packets - meterBand packets
Tags: cookie, switchid, flowid, direction, meterid

Metrics for default rule meters:

- pen.switch.flow.system.meter.packets - meterBand packets
- pen.switch.flow.system.meter.bytes - meterBand bytes
- pen.switch.flow.system.meter.bits - meterBand bits
Tags: cookie (in Hex), switchid, meterid

Closes #1573

---
This change is [Reviewable](https://reviewable.io/reviews/telstra/open-kilda/1747)

### CODE REVIEW

**Comment by nikitacherevko** [comment]
In general looks good, but there is a need to fix mapping meters and flows in order to be able to track meter per flow.

File: services/wfm/src/main/java/org/openkilda/wfm/topology/stats/bolts/CacheFilterBolt.java [hunk 1]
```diff
@@ -96,16 +98,16 @@ public void execute(Tuple tuple) {
...
                     InstallEgressFlow command = (InstallEgressFlow) data;
                     logMatchedRecord(command);
                     emit(tuple, Commands.UPDATE, command, MeasurePoint.EGRESS);
                 } else if (data instanceof InstallOneSwitchFlow) {
                     InstallOneSwitchFlow command = (InstallOneSwitchFlow) data;
                     logMatchedRecord(command);
-                    emit(tuple, Commands.UPDATE, command, MeasurePoint.INGRESS);
-                    emit(tuple, Commands.UPDATE, command, MeasurePoint.EGRESS);
+                    emit(tuple, Commands.UPDATE, command, command.getMeterId(), MeasurePoint.INGRESS);
+                    emit(tuple, Commands.UPDATE, command, command.getMeterId(), MeasurePoint.EGRESS);
```
[the remark this answers is not in the document]
Comment (niksv) [reply services/wfm/src/main/java/org/openkilda/wfm/topology/stats/bolts/CacheFilterBolt.java]: yep

File: services/wfm/src/main/java/org/openkilda/wfm/topology/stats/bolts/CacheBolt.java [hunk 2]
```diff
@@ -66,13 +71,11 @@
...
-    private TopologyContext context;
-    private OutputCollector outputCollector;
-
     /**
-     * Cookie to flow map.
+     * Cookie to flow and meter to flow maps.
      */
-    private Map<Long, CacheFlowEntry> cookieToFlow = new HashMap<>();
+    private Map<Long, FlowByCookieCacheEntry> cookieToFlow = new HashMap<>();
+    private Map<Long, FlowByMeterCacheEntry> meterToFlow = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (niksv) [reply services/wfm/src/main/java/org/openkilda/wfm/topology/stats/bolts/CacheBolt.java]: good catch!

File: services/src/floodlight-modules/src/main/java/org/openkilda/floodlight/converter/OfMeterStatsConverter.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+import org.projectfloodlight.openflow.protocol.OFMeterStatsReply;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+
+public final class OfMeterStatsConverter {
+    private static final Logger logger = LoggerFactory.getLogger(OfMeterStatsConverter.class);
+
+    private OfMeterStatsConverter() {}
```
Comment (sergii-iakovenko) [inline services/src/floodlight-modules/src/main/java/org/openkilda/floodlight/converter/OfMeterStatsConverter.java]: There's a good practice to throw a UnsupportedOperationException in private constructors of utility classes.

Comment (niksv) [reply services/src/floodlight-modules/src/main/java/org/openkilda/floodlight/converter/OfMeterStatsConverter.java]: fixed

File: services/src/messaging/src/main/java/org/openkilda/messaging/info/stats/MeterStatsReply.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+package org.openkilda.messaging.info.stats;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.EqualsAndHashCode;
+
+import java.io.Serializable;
+import java.util.List;
+
+@EqualsAndHashCode
```
[the remark this answers is not in the document]
Comment (niksv) [reply services/src/messaging/src/main/java/org/openkilda/messaging/info/stats/MeterStatsReply.java]: fixed

File: services/wfm/src/main/java/org/openkilda/wfm/topology/stats/metrics/MeterStatsMetricGenBolt.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+        Map<Pair<SwitchId, Long>, FlowByMeterCacheEntry> flowCache =
+                (Map<Pair<SwitchId, Long>, FlowByMeterCacheEntry>) input.getValueByField(METER_CACHE_FIELD);
+
+        if (!Destination.WFM_STATS.equals(message.getDestination())) {
+            collector.ack(input);
+            return;
+        }
+
+        logger.debug("Meter stats message: {}={}, component={}, stream={}",
+                CORRELATION_ID, message.getCorrelationId(), componentId,
```
[the remark this answers is not in the document]
Comment (niksv) [reply services/wfm/src/main/java/org/openkilda/wfm/topology/stats/metrics/MeterStatsMetricGenBolt.java]: ok

**Comment by sergii-iakovenko** [comment]
Please, rebase or fix merge conflicts.

**Pull request merged**